src/sidebar/reducers/annotations.js

```diff
@@ -173,17 +173,16 @@ var update = {
 
   UPDATE_ANCHOR_STATUS: function (state, action) {
     var annotations = state.annotations.map(function (annot) {
-      var match = (annot.id && annot.id === action.id) ||
-                  (annot.$tag && annot.$tag === action.tag);
-      if (match) {
-        return Object.assign({}, annot, {
-          $anchorTimeout: action.anchorTimeout || annot.$anchorTimeout,
-          $orphan: action.isOrphan,
-          $tag: action.tag,
-        });
-      } else {
+      if (!action.statusUpdates.hasOwnProperty(annot.$tag)) {
```

Comment:
Ok, so here we're looping over _all_ of the annotations in the sidebar (I guess this is why it's slow, to call this function too often) and for each annotation, _if_ the annotation is in the status updates, we proceed

Reply:
> Ok, so here we're looping over all of the annotations in the sidebar (I guess this is why it's slow, to call this function too often)

The current data structure that stores annotations in the sidebar is a list. This is indeed certainly suboptimal for certain operations. In practice however, it wasn't the (main) issue here from the profiling I looked at.

Rather the problem is that for every action that goes through the store, there are a bunch of processing steps that get triggered. These include Angular digest cycles (though these are debounced using `$applyAsync`) to trigger UI updates, plus any subscribers to state changes. Collectively this all adds up. It can be optimized, but batching updates is always going to be more efficient.